controllers/openshift_controller.go

```diff
@@ -941,47 +957,107 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigInstallRequest() (ctrl.
+			}
+		}
+		if labelingChanged {
+			r.Log.Info("node labels updated")
+			// IsInProgress == ConditionFalse would be more direct and clear but it doesn't handle
+			// the possibility that IsInProgress == "" which actually comes up right after
+			// KataConfig is created.
+			if r.kataConfig.Status.InstallationStatus.IsInProgress != corev1.ConditionTrue {
+				r.Log.Info("Starting to wait for MCO to start")
+				r.kataConfig.Status.WaitingForMcoToStart = true
+			} else {
+				r.Log.Info("installation already in progress", "IsInProgress", r.kataConfig.Status.InstallationStatus.IsInProgress)
+			}
+		}
 
 		// Create kata-oc only if it doesn't exist
-		foundMcp := &mcfgv1.MachineConfigPool{}
-		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: machinePool}, foundMcp)
+		mcp := &mcfgv1.MachineConfigPool{}
+		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: machinePool}, mcp)
 		if err != nil && k8serrors.IsNotFound(err) {
 			r.Log.Info("Creating a new MachineConfigPool ", "machinePool", machinePool)
+			mcp = r.newMCPforCR()
 			err = r.Client.Create(context.TODO(), mcp)
 			if err != nil {
 				r.Log.Error(err, "Error in creating new MachineConfigPool ", "machinePool", machinePool)
 				return ctrl.Result{}, err
 			}
-			// mcp created successfully - requeue to check the status later
-			return ctrl.Result{Requeue: true, RequeueAfter: 20 * time.Second}, nil
+			// Don't requeue - the MCP creation will send us
+			// a reconcile request via our MCP watching so we should be
+			// guaranteed to run again at due time even without requeueing.
 		} else if err != nil {
 			r.Log.Error(err, "Error in retreiving MachineConfigPool ", "machinePool", machinePool)
 			return ctrl.Result{}, err
 		}
+	} else {
+		if wasMcJustCreated {
+			r.kataConfig.Status.WaitingForMcoToStart = true
+		}
+	}
 
-		// Wait till MCP is ready
-		if foundMcp.Status.MachineCount == 0 {
-			r.Log.Info("Waiting till MachineConfigPool is initialized ", "machinePool", machinePool)
-			return ctrl.Result{Requeue: true, RequeueAfter: 15 * time.Second}, nil
+	isMcpUpdating := func(mcpName string) bool {
+		mcp := &mcfgv1.MachineConfigPool{}
+		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: mcpName}, mcp)
+		if err != nil {
+			r.Log.Info("Getting MachineConfigPool failed ", "machinePool", mcpName, "err", err)
+			return false
 		}
+		return mcfgv1.IsMachineConfigPoolConditionTrue(mcp.Status.Conditions, mcfgv1.MachineConfigPoolUpdating)
 	}
 
-	doReconcile, err, isMcCreated := r.createExtensionMc(machinePool)
-	if isMcCreated {
-		return doReconcile, err
+	isKataMcpUpdating := isMcpUpdating(machinePool)
+	r.Log.Info("MCP updating state", "MCP name", machinePool, "is updating", isKataMcpUpdating)
+	if isKataMcpUpdating {
+		r.kataConfig.Status.InstallationStatus.IsInProgress = corev1.ConditionTrue
+	}
+	isMcoUpdating := isKataMcpUpdating
+	if !isConvergedCluster {
+		isWorkerUpdating := isMcpUpdating("worker")
```

Comment:
I don't quite get why we care for the MCP being updating.

Reply:
Remember how when workers migrate from a source to a target pool, the source pool is drained instantly (without ever entering the Updating state) but the target pool goes into Updating and is joined by the migrating workers one by one over time?  From this it follows that if we want to know when the MCO is finished we need to keep an eye on the _target_ pool.

Now the target pool might be "kata-oc" for nodes that are having kata installed on them and are thus migrating from "worker" to "kata-oc", or "worker" for nodes that are having kata uninstalled from them and are migrating back to "worker".  So we know the MCO is done only when both potential target MCPs, "kata-oc" and "worker", are settled.

Reply:
@pmores What if some unrelated operation causes "worker" to update, e.g. something else creating machine configs ? Could this prevent to reach the MCO is done state ?

Reply:
BTW the original code didn't need to wait for "worker" because it only handled the initial installation where nodes can only go from "worker" to "kata-oc" by definition.  However, since we've been supporting a subsequent update of kata installation on cluster, we need to handle nodes returning back to "worker" as well.

Reply:
We cannot figure out _why_ an MCP is Updating, at least not easily if at all.  But we can still assume that the MCP should settle, finally.

So, let's say that user creates a KataConfig to put kata on a half of their workers, and they create another MCP and put the other half of their nodes in it.  Now if they if they remove some node(s) from the "kata-oc" pool and while this is running, they, say, delete their other custom MCP, nodes from both MCPs will be rejoining "worker" which will not settle until all nodes migrating to it are updated.  It will still settle although it's true that we might waiting a bit longer than necessary.

If that becomes a problem I believe we could just make the `isMcpUpdating()` function more precise (like, iterating over nodes belonging to that MCP and checking for each individually if it's Updated at the right rendered-* MC).  I also believe though that doing so at this point would smell like overengineering.

Reply:
Please add a comment to state that : `isMcpUpdating()` might catch unrelated changes, improve its precision if it becomes a problem.